backend/core/src/main/java/org/sonarsource/sonarlint/core/tracking/TrackingService.java

```diff
@@ -0,0 +1,304 @@
+
+  public TrackingService(SonarLintRpcClient client, ConfigurationRepository configurationRepository, SonarProjectBranchTrackingService branchTrackingService,
+    PathTranslationService pathTranslationService, FindingReportingService reportingService, KnownFindingsStorageService knownFindingsStorageService, StorageService storageService,
+    LocalOnlyIssueRepository localOnlyIssueRepository, LocalOnlyIssueStorageService localOnlyIssueStorageService, FindingsSynchronizationService findingsSynchronizationService) {
+    this.client = client;
+    this.configurationRepository = configurationRepository;
+    this.branchTrackingService = branchTrackingService;
+    this.pathTranslationService = pathTranslationService;
+    this.reportingService = reportingService;
+    this.knownFindingsStorageService = knownFindingsStorageService;
+    this.storageService = storageService;
+    this.localOnlyIssueRepository = localOnlyIssueRepository;
+    this.localOnlyIssueStorageService = localOnlyIssueStorageService;
+    this.findingsSynchronizationService = findingsSynchronizationService;
+  }
+
+  @EventListener
+  public void onAnalysisStarted(AnalysisStartedEvent event) {
+    if (event.isTrackingEnabled()) {
+      var matchingSession = startMatchingSession(event.getConfigurationScopeId(), event.getFiles());
+      matchingSessionByAnalysisId.put(event.getAnalysisId(), matchingSession);
+    }
+  }
+
+  @EventListener
+  public void onIssueDetected(RawIssueDetectedEvent event) {
+    var analysisId = event.getAnalysisId();
+    var matchingSession = matchingSessionByAnalysisId.get(analysisId);
+    if (matchingSession == null) {
+      // an issue was detected outside any analysis, this normally shouldn't happen
+      return;
+    }
+    var detectedIssue = event.getDetectedIssue();
+    var isSupported = detectedIssue.isInFile();
+    if (isSupported) {
+      // we don't support global issues for now
+      var trackedIssue = matchingSession.matchWithKnownFinding(requireNonNull(detectedIssue.getIdeRelativePath()), detectedIssue);
+      reportingService.stream(event.getConfigurationScopeId(), analysisId, trackedIssue);
+    }
+  }
+
+  @EventListener
+  public void onAnalysisFinished(AnalysisFinishedEvent event) {
+    var analysisId = event.getAnalysisId();
+    var matchingSession = matchingSessionByAnalysisId.remove(analysisId);
+    if (matchingSession == null) {
+      // a not-started analysis finished, this normally shouldn't happen
+      return;
+    }
+    var configurationScopeId = event.getConfigurationScopeId();
+    if (event.shouldFetchServerIssues()) {
+      findingsSynchronizationService.refreshServerFindings(configurationScopeId, matchingSession.getRelativePathsInvolved());
+    }
+    var result = matchWithServerFindings(configurationScopeId, matchingSession);
+    reportingService.report(configurationScopeId, analysisId, result.issuesToReport, result.hotspotsToReport);
+  }
+
+  @NotNull
```

Comment:
Neat pick - feel free to ignore.
I've noticed inconsistency in nullability annotations. For nullable we usually use @CheckForNull from javax.annotation and for non-nullable more often I see this from JetBrains. Maybe we should go with one source of annotations.

Reply:
Good catch, I tend to use `@CheckForNull`, here the `@NotNull` probably got automatically inserted by JetBrains. The good thing is that we always return non-null, so I got rid of the annotations in this file.